File: A09 two pointer(sliding Window)/TwoSum or KSum/LeetCode 16 Sum Closet/two pointer.py
###
#Python O(n^2) time, O(1) space two-pointers solution


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:

        diff = float('inf')
        a, b, c = -1, -1, -1
        nums.sort()
        n = len(nums)
        for i in range(n):
            start, end = i+1, n-1
            while start < end:
                s = nums[i] + nums[start] + nums[end]
                if abs(s-target) < diff:
                    diff = abs(s-target)
                    a, b, c = i, start, end
                if s < target:
                    start += 1
                else:
                    end -= 1
            if diff ==0:
                return nums[a]+nums[b]+nums[c]
        return nums[a]+nums[b]+nums[c]

File: A09 two pointer(sliding Window)/TwoSum or KSum/LeetCode 16 Sum Closet/two pointer.java
//6ms
//

class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int result = 0; 
        int minGap = Integer.MAX_VALUE; 
        Arrays.sort(nums); 
        
        for (int i = 0; i < nums.length -2; i ++){
            int j = i +1; 
            int k = nums.length -1; 
            
            while (j < k){å
                int sum = nums[i] + nums[j] + nums[k]; 
                int gap = Math.abs(sum - target); 
                if(gap < minGap){
                    minGap = gap; 
                    result = sum;
                }
                
                if (sum < target){
                    j++;
                } else{
                    k--;
                }
            }
        }
        return result;
    }

}
